XFS: ensure proper alignment of extent allocation size hint

pwrobelse opened the pull request:
If the extent size is not aligned to logical block
size then XFS ignores it and returns an error from
ioctl() function call.

These patches ensure that the hint passed by a user
is properly aligned.

xemul commented:
> When it is not aligned to 4KB, then the hint seems to be ignored.

Why is that?

pwrobelse commented:
> > When it is not aligned to 4KB, then the hint seems to be ignored.
> 
> Why is that?

I am not entirely sure why 4K helps - maybe the block size would be sufficient. The code in seastar ignores the error (reactor.cc):
```
// Ignore error; may be !xfs, and just a hint anyway
if (r != -1) {
    attr.fsx_xflags |= XFS_XFLAG_EXTSIZE;
    attr.fsx_extsize = std::min(options.extent_allocation_size_hint, file_open_options::max_extent_allocation_size_hint);

      // Ignore error; may be !xfs, and just a hint anyway
      ::ioctl(fd, XFS_IOC_FSSETXATTR, &attr);
}
```

I passed the following configuration:
```
- name: unlinking
  shards: all
  type: unlink
  data_size: 21GB
  files_count: 210
  shard_info:
    parallelism: 1
    rps: 1
```

Given that the test uses 4 shards then each shard gets `21 * (2^20) / 4 = 21*(2^18)`. Then each shard splits this amount of memory to create files: `21 * (2^18) / 210 = int(26214.4) = 26214`. The given size is neither divisible by 4096 nor 512.

When I run `xfs_bmap` on the created files I got the following results.

OLD: Without alignment to 4096:
```
sudo xfs_bmap -e -l -p -v -v -v  /mnt/my_xfs/test-unlinking-shard-3-file-87
/mnt/my_xfs/test-unlinking-shard-3-file-87:
 EXT: FILE-OFFSET      BLOCK-RANGE          AG AG-OFFSET            TOTAL FLAGS
   0: [0..1535]:       9597480..9599015      0 (9597480..9599015)    1536 000000
   1: [1536..2047]:    9599528..9600039      0 (9599528..9600039)     512 000000
   2: [2048..2559]:    9600552..9601063      0 (9600552..9601063)     512 000000
   3: [2560..3583]:    9608232..9609255      0 (9608232..9609255)    1024 000000
   4: [3584..4095]:    9611304..9611815      0 (9611304..9611815)     512 000000
   5: [4096..5119]:    10704936..10705959    0 (10704936..10705959)  1024 000000
   6: [5120..9215]:    11969576..11973671    0 (11969576..11973671)  4096 000000
   7: [9216..11775]:   13556264..13558823    0 (13556264..13558823)  2560 000000
   8: [11776..12287]:  13569576..13570087    0 (13569576..13570087)   512 000000
   9: [12288..13823]:  15102504..15104039    0 (15102504..15104039)  1536 000000
  10: [13824..15871]:  16591400..16593447    0 (16591400..16593447)  2048 000000
  11: [15872..19967]:  17846824..17850919    0 (17846824..17850919)  4096 000000
  12: [19968..20479]:  19266600..19267111    0 (19266600..19267111)   512 000000
  13: [20480..23039]:  19273256..19275815    0 (19273256..19275815)  2560 000000
  14: [23040..26111]:  20744232..20747303    0 (20744232..20747303)  3072 000000
  15: [26112..26623]:  22244392..22244903    0 (22244392..22244903)   512 000000
  16: [26624..30207]:  23639080..23642663    0 (23639080..23642663)  3584 000000
  17: [30208..30719]:  23654952..23655463    0 (23654952..23655463)   512 000000
  18: [30720..32767]:  24763968..24766015    0 (24763968..24766015)  2048 000000
  19: [32768..34815]:  24777792..24779839    0 (24777792..24779839)  2048 000000
  20: [34816..36863]:  25701688..25703735    0 (25701688..25703735)  2048 000000
  21: [36864..37887]:  27680760..27681783    0 (27680760..27681783)  1024 000000
  22: [37888..39423]:  28710496..28712031    0 (28710496..28712031)  1536 000000
  23: [39424..40959]:  28718688..28720223    0 (28718688..28720223)  1536 000000
  24: [40960..41471]:  28738656..28739167    0 (28738656..28739167)   512 000000
  25: [41472..41983]:  28789344..28789855    0 (28789344..28789855)   512 000000
  26: [41984..42495]:  29991520..29992031    0 (29991520..29992031)   512 000000
  27: [42496..44031]:  30000224..30001759    0 (30000224..30001759)  1536 000000
  28: [44032..44543]:  30009952..30010463    0 (30009952..30010463)   512 000000
  29: [44544..45055]:  475409808..475410319  1 (11542616..11543127)   512 000000
  30: [45056..46079]:  476155896..476156919  1 (12288704..12289727)  1024 000000
  31: [46080..46591]:  476328464..476328975  1 (12461272..12461783)   512 000000
  32: [46592..48127]:  477000728..477002263  1 (13133536..13135071)  1536 000000
  33: [48128..48639]:  477175832..477176343  1 (13308640..13309151)   512 000000
  34: [48640..51711]:  477719072..477722143  1 (13851880..13854951)  3072 000000
  35: [51712..52223]:  477760032..477760543  1 (13892840..13893351)   512 000000
  36: [52224..52735]:  477761056..477761567  1 (13893864..13894375)   512 000000
```


NEW: With alignment to 4096:
```
sudo xfs_bmap -e -l -p -v -v -v  /mnt/my_xfs/test-unlinking-shard-3-file-87
/mnt/my_xfs/test-unlinking-shard-3-file-87:
 EXT: FILE-OFFSET      BLOCK-RANGE          AG AG-OFFSET            TOTAL FLAGS
   0: [0..52431]:      482742800..482795231  1 (18875608..18928039) 52432 000000
   1: [52432..52735]:  532762928..532763231  1 (68895736..68896039)   304 000000
 FLAG Values:
    0100000 Shared extent
    0010000 Unwritten preallocated extent
    0001000 Doesn't begin on stripe unit
    0000100 Doesn't end   on stripe unit
    0000010 Doesn't begin on stripe width
    0000001 Doesn't end   on stripe width
```

xemul commented:
OK, so here's the kernel code

```c
    if (S_ISREG(VFS_I(ip)->i_mode) && ip->i_df.if_nextents &&
        XFS_FSB_TO_B(mp, ip->i_extsize) != fa->fsx_extsize)
        return -EINVAL;

    if (fa->fsx_extsize & mp->m_blockmask)
        return -EINVAL;
```

xemul commented:
It looks like it deserves two fixes :)
1. reactor.cc code should round-up the user provided size-hint to ... I would put the magic named-constant of 128k
2. io-tester should round-up the calculated file-size to 1Mb

pwrobelse commented:
> It looks like it deserves two fixes :)
> 
>     1. reactor.cc code should round-up the user provided size-hint to ... I would put the magic named-constant of 128k
> 
>     2. io-tester should round-up the calculated file-size to 1Mb

Okay, done :+1:

xemul commented on apps/io_tester/io_tester.cc:
```diff
@@ -772,7 +772,7 @@ class unlink_class_data : public class_data {
         return max_concurrent_for_each(boost::irange(UINT64_C(0), files_count()), max_concurrency(), [this] (uint64_t file_id) {
             const auto fname = get_filename(file_id);
-            const auto fsize = _config.file_size / files_count();
+            const auto fsize = align_up<uint64_t>(_config.file_size / files_count(), 4096u);
```
Why not extent_size_hint_alignment as in previous patch?

pwrobelse replied:
Fixed

xemul commented on apps/io_tester/io_tester.cc:
```diff
@@ -774,7 +776,7 @@ class unlink_class_data : public class_data {
 
             file_open_options options;
-            options.extent_allocation_size_hint = fsize;
+            options.extent_allocation_size_hint = align_up<uint64_t>(fsize, extent_size_hint_alignment);
```
It's not extent allocation that should be aligned by io-tester, but rather the file size itself

pwrobelse replied:
Sure - fixed for all job types in io_tester :+1:

xemul commented on src/core/reactor.cc:
```diff
@@ -1905,6 +1905,13 @@ reactor::open_file_dma(std::string_view nameref, open_flags flags, file_open_opt
+                    constexpr uint32_t min_hint_alignment_mask{min_extent_size_hint_alignment - 1u};
+
+                    if (attr.fsx_extsize & min_hint_alignment_mask) {
+                        attr.fsx_extsize = align_up<uint32_t>(attr.fsx_extsize, min_extent_size_hint_alignment);
```
Doesn't `align_up`  already leaves aligned values as is, so the if above is excessive?

pwrobelse replied:
Yes, that's true. It is a leftover from testing - in the beginning I added a warning here to check how it behaves.

pwrobelse replied:
Fixed

pwrobelse commented:
Description of changes between patches:
- aligned the expected file size when it is read from YAML file in io_tester.cc
- removed redundant if statement from reactor.cc
- updated the documentation of io_tester to inform users about the alignment of file size
- adjusted the function that writes data to newly created file to ensure appropriate calculation of required buffers count - previously when file size was divisible by write buffer size, then we did one unneeded write that caused creation of one more extent for given file

xemul commented on src/core/reactor.cc:
```diff
@@ -1905,6 +1905,9 @@ reactor::open_file_dma(std::string_view nameref, open_flags flags, file_open_opt
                     attr.fsx_extsize = std::min(options.extent_allocation_size_hint,
                                         file_open_options::max_extent_allocation_size_hint);
 
+                    constexpr uint32_t min_extent_size_hint_alignment{128u << 10}; // 128KB
```
Please, put it next to `file_open_options::max_extent_allocation_size_hint` with a comment that XFS rejects hints that are not aligned, so we try to please it with the best guess

pwrobelse replied:
Sure, done
